File: buzzvil-sdk-v6-ios/buzzvil-sdk-ios-objc/EntryPoint/EntryPointViewController.m
#import "EntryPointViewController.h"
@import BuzzvilSDK;

@interface EntryPointViewController ()

@property (nonatomic, strong) UIStackView *stackView;
@property (nonatomic, strong) UIButton *showFabButton;
@property (nonatomic, strong) UIButton *showBannerButton;
@property (nonatomic, strong) UIButton *showPopupButton;
@property (nonatomic, strong) UIButton *showBottomSheetButton;
@property (nonatomic, strong) UIButton *customEntryButton;
@property (nonatomic, strong) BuzzEntryPointBannerView *entryBannerView;
@property (nonatomic, strong) UIView *spacingView;
@property (nonatomic, strong) UIActivityIndicatorView *activityIndicator;
@property (nonatomic, assign) BOOL isFabPresented;

@end

@implementation EntryPointViewController

- (void)viewDidLoad {
  [super viewDidLoad];
  [self setup];
  [self loadEntryPoints];
}

- (void)setup {
  self.view.backgroundColor = [UIColor systemBackgroundColor];
  
  self.stackView = [[UIStackView alloc] initWithFrame:CGRectZero];
  self.stackView.axis = UILayoutConstraintAxisVertical;
  self.stackView.spacing = 12;
  self.stackView.distribution = UIStackViewDistributionFill;
  self.stackView.translatesAutoresizingMaskIntoConstraints = NO;
  
  self.showFabButton = [self createButtonWithTitle:@"Show FAB" action:@selector(showFabButtonTapped)];
  self.showBannerButton = [self createButtonWithTitle:@"Show Banner" action:@selector(showBannerButtonTapped)];
  self.showPopupButton = [self createButtonWithTitle:@"Show Popup" action:@selector(showPopupButtonTapped)];
  self.showBottomSheetButton = [self createButtonWithTitle:@"Show BottomSheet" action:@selector(showBottomSheetButtonTapped)];
  self.customEntryButton = [self createButtonWithTitle:@"Custom Entry" action:@selector(customEntryButtonTapped)];
  
  self.entryBannerView = [[BuzzEntryPointBannerView alloc] initWithFrame:CGRectZero];
  self.entryBannerView.hidden = YES;
  self.entryBannerView.translatesAutoresizingMaskIntoConstraints = NO;
  
  self.spacingView = [[UIView alloc] initWithFrame:CGRectZero];
  [self.spacingView setContentHuggingPriority:UILayoutPriorityDefaultLow forAxis:UILayoutConstraintAxisVertical];
  [self.spacingView setContentCompressionResistancePriority:UILayoutPriorityDefaultLow forAxis:UILayoutConstraintAxisVertical];
  
  self.activityIndicator = [[UIActivityIndicatorView alloc] initWithActivityIndicatorStyle:UIActivityIndicatorViewStyleMedium];
  self.activityIndicator.hidesWhenStopped = YES;
  self.activityIndicator.translatesAutoresizingMaskIntoConstraints = NO;
  
  [self.view addSubview:self.stackView];
  [self.stackView addArrangedSubview:self.showFabButton];
  [self.stackView addArrangedSubview:self.showBannerButton];
  [self.stackView addArrangedSubview:self.showPopupButton];
  [self.stackView addArrangedSubview:self.showBottomSheetButton];
  [self.stackView addArrangedSubview:self.customEntryButton];
  [self.stackView addArrangedSubview:self.entryBannerView];
  [self.stackView addArrangedSubview:self.spacingView];
  [self.view addSubview:self.activityIndicator];
  
  NSArray<UIButton *> *buttons = @[
    self.showFabButton,
    self.showBannerButton,
    self.showPopupButton,
    self.showBottomSheetButton,
    self.customEntryButton
  ];
  
  for (UIButton *button in buttons) {
    [self setupAppearanceOfButton:button];
  }
  
  [NSLayoutConstraint activateConstraints:@[
    [self.stackView.topAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.topAnchor],
    [self.stackView.leadingAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.leadingAnchor],
    [self.stackView.trailingAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.trailingAnchor],
    [self.stackView.bottomAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.bottomAnchor constant:-100],
    
    [self.entryBannerView.heightAnchor constraintEqualToAnchor:self.entryBannerView.widthAnchor multiplier:(112.0/351.0)],
    
    [self.activityIndicator.topAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.topAnchor],
    [self.activityIndicator.leadingAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.leadingAnchor],
    [self.activityIndicator.trailingAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.trailingAnchor],
    [self.activityIndicator.bottomAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.bottomAnchor],
  ]];
}

- (UIButton *)createButtonWithTitle:(NSString *)title action:(SEL)selector {
  UIButton *button = [[UIButton alloc] initWithFrame:CGRectZero];
  [button setTitle:title forState:UIControlStateNormal];
  [button addTarget:self action:selector forControlEvents:UIControlEventTouchUpInside];
  button.translatesAutoresizingMaskIntoConstraints = NO;
  return button;
}

- (void)setupAppearanceOfButton:(UIButton *)button {
  button.backgroundColor = [UIColor systemBlueColor];
  button.layer.cornerRadius = 8;
  [button setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
  button.titleLabel.font = [UIFont systemFontOfSize:16];
  [NSLayoutConstraint activateConstraints:@[
    [button.heightAnchor constraintEqualToConstant:48]
  ]];
}

- (void)loadEntryPoints {
  [self.activityIndicator startAnimating];
  
  __weak typeof(self) weakSelf = self;
  [[BuzzEntryPoint shared] loadOnSuccess:^(NSArray * _Nonnull entryPoints) {
    __strong typeof(self) strongSelf = weakSelf;
    [strongSelf.activityIndicator stopAnimating];
  } onFailure:^(NSError * _Nonnull error) {
    __strong typeof(self) strongSelf = weakSelf;
    [strongSelf.activityIndicator stopAnimating];
  }];
}

- (void)showFabButtonTapped {
  if (![[BuzzEntryPoint shared] canShowWithType:BuzzEntryPointTypeFab] || self.isFabPresented) {
    return;
  }
  [[BuzzEntryPoint shared] showFabOn:self];
  self.isFabPresented = YES;
}

- (void)showBannerButtonTapped {
  if (![[BuzzEntryPoint shared] canShowWithType:BuzzEntryPointTypeBanner]) {
    return;
  }
  self.entryBannerView.hidden = NO;
  [[BuzzEntryPoint shared] showBannerOn:self.entryBannerView];
}

- (void)showPopupButtonTapped {
  if (![[BuzzEntryPoint shared] canShowWithType:BuzzEntryPointTypePopup]) {
    return;
  }
  [[BuzzEntryPoint shared] showPopupOn:self];
}

- (void)showBottomSheetButtonTapped {
  if (![[BuzzEntryPoint shared] canShowWithType:BuzzEntryPointTypeBottomSheet]) {
    return;
  }
  [[BuzzEntryPoint shared] showBottomSheetOn:self];
}

- (void)customEntryButtonTapped {
  [[BuzzEntryPoint shared] customEntryPointClickedWithIdentifier:@"YOUR_CUSTOM_ENTRY_IDENTIFIER" on:self];
}

@end

File: buzzvil-sdk-v6-ios/buzzvil-sdk-ios-swift/EntryPoint/EntryPointViewController.swift
import UIKit
import BuzzvilSDK

final class EntryPointViewController: UIViewController {
  private lazy var stackView: UIStackView = {
    let stackView = UIStackView(frame: .zero)
    stackView.axis = .vertical
    stackView.spacing = 12
    stackView.distribution = .fill
    return stackView
  }()
  
  private lazy var showFabButton: UIButton = {
    let button = UIButton(frame: .zero)
    button.setTitle("Show FAB", for: .normal)
    button.addTarget(self, action: #selector(showFabButtonTapped), for: .touchUpInside)
    return button
  }()
  
  private lazy var showBannerButton: UIButton = {
    let button = UIButton(frame: .zero)
    button.setTitle("Show Banner", for: .normal)
    button.addTarget(self, action: #selector(showBannerButtonTapped), for: .touchUpInside)
    return button
  }()
  
  private lazy var showPopupButton: UIButton = {
    let button = UIButton(frame: .zero)
    button.setTitle("Show Popup", for: .normal)
    button.addTarget(self, action: #selector(showPopupButtonTapped), for: .touchUpInside)
    return button
  }()
  
  private lazy var showBottomSheetButton: UIButton = {
    let button = UIButton(frame: .zero)
    button.setTitle("Show BottomSheet", for: .normal)
    button.addTarget(self, action: #selector(showBottomSheetButtonTapped), for: .touchUpInside)
    return button
  }()
  
  private lazy var customEntryButton: UIButton = {
    let button = UIButton(frame: .zero)
    button.setTitle("Custom Entry", for: .normal)
    button.addTarget(self, action: #selector(customEntryButtonTapped), for: .touchUpInside)
    return button
  }()
  
  private lazy var entryBannerView: BuzzEntryPointBannerView = {
    let view = BuzzEntryPointBannerView(frame: .zero)
    view.isHidden = true
    return view
  }()
  
  private lazy var spacingView: UIView = {
    return UIView(frame: .zero)
  }()
  
  private lazy var activityIndicator: UIActivityIndicatorView = {
    let indicatorView = UIActivityIndicatorView(style: .medium)
    indicatorView.hidesWhenStopped = true
    return indicatorView
  }()
  
  private var isFabPresented = false
  
  override func viewDidLoad() {
    super.viewDidLoad()
    setup()
    loadEntryPoints()
  }
  
  private func setup() {
    view.backgroundColor = .systemBackground
    
    view.addSubview(stackView)
    stackView.addArrangedSubview(showFabButton)
    stackView.addArrangedSubview(showBannerButton)
    stackView.addArrangedSubview(showPopupButton)
    stackView.addArrangedSubview(showBottomSheetButton)
    stackView.addArrangedSubview(customEntryButton)
    stackView.addArrangedSubview(entryBannerView)
    stackView.addArrangedSubview(spacingView)
    view.addSubview(activityIndicator)
    
    let buttonsToSetupAppearance = [
      showFabButton,
      showBannerButton,
      showPopupButton,
      showBottomSheetButton,
      customEntryButton
    ]
    
    buttonsToSetupAppearance.forEach { button in
      setupAppearanceOfButton(button: button)
    }
    
    stackView.translatesAutoresizingMaskIntoConstraints = false
    NSLayoutConstraint.activate([
      stackView.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor),
      stackView.leadingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.leadingAnchor),
      stackView.trailingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.trailingAnchor),
      stackView.bottomAnchor.constraint(equalTo: view.safeAreaLayoutGuide.bottomAnchor, constant: -100),
    ])
    
    entryBannerView.translatesAutoresizingMaskIntoConstraints = false
    NSLayoutConstraint.activate([
      entryBannerView.heightAnchor.constraint(equalTo: entryBannerView.widthAnchor, multiplier: 112.0/351.0),
    ])
    
    spacingView.setContentHuggingPriority(.defaultLow, for: .vertical)
    spacingView.setContentCompressionResistancePriority(.defaultLow, for: .vertical)
    
    activityIndicator.translatesAutoresizingMaskIntoConstraints = false
    NSLayoutConstraint.activate([
      activityIndicator.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor),
      activityIndicator.leadingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.leadingAnchor),
      activityIndicator.trailingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.trailingAnchor),
      activityIndicator.bottomAnchor.constraint(equalTo: view.safeAreaLayoutGuide.bottomAnchor),
    ])
  }
  
  private func setupAppearanceOfButton(button: UIButton) {
    button.backgroundColor = .systemBlue
    button.layer.cornerRadius = 8
    button.setTitleColor(.white, for: .normal)
    button.titleLabel?.font = UIFont.systemFont(ofSize: 16)
    button.translatesAutoresizingMaskIntoConstraints = false
    NSLayoutConstraint.activate([
      button.heightAnchor.constraint(equalToConstant: 48)
    ])
  }
  
  private func loadEntryPoints() {
    activityIndicator.startAnimating()
    
    BuzzEntryPoint.shared.load(
      onSuccess: { [weak self] _ in
        self?.activityIndicator.stopAnimating()
      },
      onFailure: { [weak self] _ in
        self?.activityIndicator.stopAnimating()
      }
    )
  }
  
  @objc
  private func showFabButtonTapped() {
    guard BuzzEntryPoint.shared.canShow(type: .fab) else { return }
    guard !isFabPresented else { return }
    BuzzEntryPoint.shared.showFab(on: self)
    isFabPresented = true
  }
  
  @objc
  private func showBannerButtonTapped() {
    guard BuzzEntryPoint.shared.canShow(type: .banner) else { return }
    entryBannerView.isHidden = false
    BuzzEntryPoint.shared.showBanner(on: entryBannerView)
  }
  
  @objc
  private func showPopupButtonTapped() {
    guard BuzzEntryPoint.shared.canShow(type: .popup) else { return }
    BuzzEntryPoint.shared.showPopup(on: self)
  }
  
  @objc
  private func showBottomSheetButtonTapped() {
    guard BuzzEntryPoint.shared.canShow(type: .bottomSheet) else { return }
    BuzzEntryPoint.shared.showBottomSheet(on: self)
  }
  
  @objc
  private func customEntryButtonTapped() {
    BuzzEntryPoint.shared.customEntryPointClicked(identifier: "YOUR_CUSTOM_ENTRY_IDENTIFIER", on: self)
  }
}
